In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
data = pd.read_excel('Online retail.xlsx', sheet_name='Sheet1')
data

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                                burgers,meatballs,eggs                                                                                                                                                                             
1                                               chutney                                                                                                                                                                             
2                                        turkey,avocado                                                                                                                                                                             
3     mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                        low fat yogurt                                                                                                                                                                             
...                                                 ...                                                                                                                                                                             
7495                      butter,light mayo,fresh bread                                                                                                                                                                             
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                             
7497                                            chicken                                                                                                                                                                             
7498                                 escalope,green tea                                                                                                                                                                             
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt                                                                                                                                                                             

[7500 rows x 1 columns]

In [ ]:
trans = data.iloc[:, 0].str.split(',').tolist()
trans[:10]

[['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
 ['low fat yogurt'],
 ['whole wheat pasta', 'french fries'],
 ['soup', 'light cream', 'shallot'],
 ['frozen vegetables', 'spaghetti', 'green tea'],
 ['french fries'],
 ['eggs', 'pet food']]

In [ ]:
item_counts = data.apply(pd.Series.value_counts).sum(axis=1).sort_values(ascending=False)
item_counts.head(10)

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
cookies           223
french fries      132
escalope          101
mineral water      93
eggs               90
chocolate          73
green tea          72
spaghetti          52
champagne          45
low fat yogurt     45
dtype: int64

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
trans_encoder = TransactionEncoder()
te_ary = trans_encoder.fit(trans).transform(trans)
df_encoded = pd.DataFrame(te_ary, columns=trans_encoder.columns_)
df_encoded.head()

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False    False              False      False    False        False   
1       False    False              False      False    False        False   
2       False    False              False      False     True        False   
3       False    False              False      False    False        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False           False   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...    True           False   
3  False           False      False        False  ...   False           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False             False              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False  False        False     False  
1             False  False        False     False  
2             False  False        False     False  
3              True  False        False     False  
4             False  False        False     False  

[5 rows x 120 columns]

In [ ]:
df_encoded.shape

(7500, 120)

In [ ]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [ ]:
frequent_itemsets = apriori(df_encoded, min_support=0.01, use_colnames=True)
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                                 itemsets
0    0.020267                                (almonds)
1    0.033200                                (avocado)
2    0.010800                         (barbecue sauce)
3    0.014267                              (black tea)
4    0.011467                             (body spray)
..        ...                                      ...
254  0.011067       (ground beef, milk, mineral water)
255  0.017067  (ground beef, spaghetti, mineral water)
256  0.015733         (spaghetti, milk, mineral water)
257  0.010267    (olive oil, spaghetti, mineral water)
258  0.011467     (spaghetti, pancakes, mineral water)

[259 rows x 2 columns]

**APPLYING APRIORI**

In [ ]:
apriori_rules = association_rules(frequent_itemsets, metric="lift", min_threshold=0.7)
apriori_rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                 consequents  \
0                     (avocado)             (mineral water)   
1               (mineral water)                   (avocado)   
2                        (cake)                   (burgers)   
3                     (burgers)                      (cake)   
4                     (burgers)                 (chocolate)   
..                          ...                         ...   
431  (mineral water, spaghetti)                  (pancakes)   
432   (mineral water, pancakes)                 (spaghetti)   
433                 (spaghetti)   (mineral water, pancakes)   
434                  (pancakes)  (mineral water, spaghetti)   
435             (mineral water)       (pancakes, spaghetti)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.033200            0.238267  0.011467    0.345382  1.449559   
1              0.238267            0.033200  0.011467    0.048125  1.449559   
2              0.081067            0.087200  0.011467    0.141447  1.622103   
3              0.087200            0.081067  0.011467    0.131498  1.622103   
4              0.087200            0.163867  0.017067    0.195719  1.194377   
..                  ...                 ...       ...         ...       ...   
431            0.059733            0.095067  0.011467    0.191964  2.019260   
432            0.033733            0.174133  0.011467    0.339921  1.952073   
433            0.174133            0.033733  0.011467    0.065850  1.952073   
434            0.095067            0.059733  0.011467    0.120617  2.019260   
435            0.238267            0.025200  0.011467    0.048125  1.909736   

     leverage  conviction  zhangs_metric  
0    0.003556    1.163629       0.320785  
1    0.003556    1.015680       0.407144  
2    0.004398    1.063185       0.417349  
3    0.004398    1.058068       0.420154  
4    0.002777    1.039603       0.178291  
..        ...         ...            ...  
431  0.005788    1.119917       0.536836  
432  0.005593    1.251163       0.504751  
433  0.005593    1.034381       0.590560  
434  0.005788    1.069235       0.557797  
435  0.005462    1.024084       0.625373  

[436 rows x 10 columns]

**FILTERING AND SORTING RULES**

In [ ]:
apriori_rules  = apriori_rules .sort_values('lift', ascending=False)
apriori_rules  = apriori_rules [(apriori_rules ['confidence'] >= 0.4) & (apriori_rules ['lift'] > 0.7)]
apriori_rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents      consequents  antecedent support  \
413    (ground beef, mineral water)      (spaghetti)            0.040933   
370             (ground beef, eggs)  (mineral water)            0.020000   
406             (ground beef, milk)  (mineral water)            0.022000   
346        (ground beef, chocolate)  (mineral water)            0.023067   
394       (frozen vegetables, milk)  (mineral water)            0.023600   
300                          (soup)  (mineral water)            0.050533   
430           (pancakes, spaghetti)  (mineral water)            0.025200   
424          (olive oil, spaghetti)  (mineral water)            0.022933   
418               (milk, spaghetti)  (mineral water)            0.035467   
352               (chocolate, milk)  (mineral water)            0.032133   
412        (ground beef, spaghetti)  (mineral water)            0.039200   
400  (frozen vegetables, spaghetti)  (mineral water)            0.027867   
376                    (eggs, milk)  (mineral water)            0.030800   
290                     (olive oil)  (mineral water)            0.065733   
246                   (ground beef)  (mineral water)            0.098267   
334               (eggs, chocolate)  (mineral water)            0.033200   
364          (chocolate, spaghetti)  (mineral water)            0.039200   

     consequent support   support  confidence      lift  leverage  conviction  \
413            0.174133  0.017067    0.416938  2.394361  0.009939    1.416431   
370            0.238267  0.010133    0.506667  2.126469  0.005368    1.544054   
406            0.238267  0.011067    0.503030  2.111207  0.005825    1.532756   
346            0.238267  0.010933    0.473988  1.989319  0.005437    1.448130   
394            0.238267  0.011067    0.468927  1.968075  0.005444    1.434328   
300            0.238267  0.023067    0.456464  1.915771  0.011026    1.401441   
430            0.238267  0.011467    0.455026  1.909736  0.005462    1.397744   
424            0.238267  0.010267    0.447674  1.878880  0.004802    1.379138   
418            0.238267  0.015733    0.443609  1.861817  0.007283    1.369061   
352            0.238267  0.014000    0.435685  1.828559  0.006344    1.349836   
412            0.238267  0.017067    0.435374  1.827256  0.007727    1.349094   
400            0.238267  0.012000    0.430622  1.807311  0.005360    1.337834   
376            0.238267  0.013067    0.424242  1.780536  0.005728    1.323011   
290            0.238267  0.027467    0.417850  1.753707  0.011805    1.308483   
246            0.238267  0.040933    0.416554  1.748266  0.017520    1.305576   
334            0.238267  0.013467    0.405622  1.702389  0.005556    1.281565   
364            0.238267  0.015867    0.404762  1.698777  0.006527    1.279712   

     zhangs_metric  
413       0.607207  
370       0.540548  
406       0.538177  
346       0.509058  
394       0.503778  
300       0.503458  
430       0.488682  
424       0.478747  
418       0.479911  
352       0.468165  
412       0.471202  
400       0.459496  
376       0.452302  
290       0.460018  
246       0.474647  
334       0.426758  
364       0.428124

**TOP 10 RESULTS**

In [ ]:
apriori_rules.head(10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents      consequents  antecedent support  \
413  (ground beef, mineral water)      (spaghetti)            0.040933   
370           (ground beef, eggs)  (mineral water)            0.020000   
406           (ground beef, milk)  (mineral water)            0.022000   
346      (ground beef, chocolate)  (mineral water)            0.023067   
394     (frozen vegetables, milk)  (mineral water)            0.023600   
300                        (soup)  (mineral water)            0.050533   
430         (pancakes, spaghetti)  (mineral water)            0.025200   
424        (olive oil, spaghetti)  (mineral water)            0.022933   
418             (milk, spaghetti)  (mineral water)            0.035467   
352             (chocolate, milk)  (mineral water)            0.032133   

     consequent support   support  confidence      lift  leverage  conviction  \
413            0.174133  0.017067    0.416938  2.394361  0.009939    1.416431   
370            0.238267  0.010133    0.506667  2.126469  0.005368    1.544054   
406            0.238267  0.011067    0.503030  2.111207  0.005825    1.532756   
346            0.238267  0.010933    0.473988  1.989319  0.005437    1.448130   
394            0.238267  0.011067    0.468927  1.968075  0.005444    1.434328   
300            0.238267  0.023067    0.456464  1.915771  0.011026    1.401441   
430            0.238267  0.011467    0.455026  1.909736  0.005462    1.397744   
424            0.238267  0.010267    0.447674  1.878880  0.004802    1.379138   
418            0.238267  0.015733    0.443609  1.861817  0.007283    1.369061   
352            0.238267  0.014000    0.435685  1.828559  0.006344    1.349836   

     zhangs_metric  
413       0.607207  
370       0.540548  
406       0.538177  
346       0.509058  
394       0.503778  
300       0.503458  
430       0.488682  
424       0.478747  
418       0.479911  
352       0.468165

In [ ]:
print(f"\nTotal number of rules: {len(apriori_rules )}")


Total number of rules: 17


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


# Analysis and Interpretation:

**The top rule suggests that customers who buy '(ground beef, mineral water)' are about 2.39 times more likely to also buy '(spaghetti)' than the average customer.**

**This rule has a confidence of about 41.3%, meaning that 41.3% of the transactions containing 'ground beef, mineral water' also contain 'spaghetti'.**




**Mineral water is the most frequently purchased item, followed by eggs, spaghetti, and french fries. This suggests that hydration and convenience foods are significant in customer preferences.**


# **1. What is lift and why is it important in Association rules?**
Lift is used to compare the strength of the association between two items to the expected strength of the association if the items were independent. A lift value greater than 1 indicates that the association between two items is stronger than expected based on the frequency of the individual items.

The lift value is an important measure in association rule mining because it provides a way to distinguish between significant and insignificant association rules. A high lift value indicates that the association rule is more significant, as it suggests that the two items are highly dependent on each other.

# **2. What is support and Confidence. How do you calculate them?**

In data mining, support refers to the relative frequency of an item set in a dataset. For example, if an itemset occurs in 5% of the transactions in a dataset, it has a support of 5%. Support is often used as a threshold for identifying frequent item sets in a dataset, which can be used to generate association rules.


Confidence is a measure of the reliability or support for a given association rule. It is defined as the proportion of cases in which the association rule holds true, or in other words, the percentage of times that the items in the antecedent appear in the same transaction as the items in the consequence.
Confidence is a measure of the likelihood that an itemset will appear if another itemset appears. For example, suppose we have a dataset of 1000 transactions, and the itemset appears in 100 of those transactions. The itemset appears in 200 of those transactions.

# **3. What are some limitations or challenges of Association rules mining?**

Challenge 1: Choosing the right metrics

Challenge 2: Setting the appropriate thresholds

Challenge 3: Handling noise and outliers

Challenge 4: Dealing with missing values

Challenge 5: Managing the complexity and scalability
